### PR TITLE
 vulnerability fixes for Sol003

### LINKED ISSUE #23: sol003-lifecycle-driver-vulnerabilities-issues
We have found the sol003-lifecycledriver-vulnerabilities given below CVE's.

**CVE-2021-33037** 
                 Apache Tomcat 10.0.0-M1 to 10.0.6, 9.0.0.M1 to 9.0.46 and 8.5.0 to 8.5.66 did not correctly parse the HTTP transfer-encoding request header in some.

**CVE-2021-30640**
                 A vulnerability in the JNDI Realm of Apache Tomcat allows an attacker to authenticate using variations of a valid user name and/or to bypass some of the protection provided by the LockOut Realm. This issue affects Apache Tomcat 10.0.0-M1 to 10.0.5; 9.0.0.M1 to 9.0.45; 8.5.0 to 8.5.65.

**CVE-2021-41079**
                 Apache Tomcat 8.5.0 to 8.5.63, 9.0.0-M1 to 9.0.43 and 10.0.0-M1 to 10.0.2 did not properly validate incoming TLS packets. When Tomcat was configured to use NIO+OpenSSL or NIO2+OpenSSL for TLS, a specially crafted packet could be used to trigger an infinite loop resulting in a denial of service.

**CVE-2021-42340​**
                 The fix for bug 63362 present in Apache Tomcat 10.1.0-M1 to 10.1.0-M5, 10.0.0-M1 to 10.0.11, 9.0.40 to 9.0.53 and 8.5.60 to 8.5.71 introduced a memory leak. The object introduced to collect metrics for HTTP upgrade connections was not released for WebSocket connections once the connection was closed. This created a memory leak that, over time, could lead to a denial of service via an OutOfMemoryError.

**CVE-2021-29425**
                  In Apache Commons IO before 2.7, When invoking the method FileNameUtils.normalize with an improper input string, like '//../foo', or '\\..\\foo', the result would be the same value, thus possibly providing access to files in the parent directory, but not further above (thus 'limited' path traversal), if the calling code would use the result to construct a path value.

**CVE-2020-9488**
                 Improper validation of certificate with host mismatch in Apache Log4j SMTP appender. This could allow an SMTPS connection to be intercepted by a man-in-the-middle attack which could leak any log messages sent through that appender.

**PRISMA-2021-0055**
                Versions <1.13 of this package are vulnerable to Information Exposure. When there is no byte array value that can be encoded into a string, the Base32 implementation does not reject it, and instead decodes it into an arbitrary value which can be re-encoded again using the same implementation. This allows for information exposure exploits such as tunneling additional information via seemingly valid base 32 strings. 

**CVE-2019-17495**
                 A Cascading Style Sheets (CSS) injection vulnerability in Swagger UI before 3.23.11 allows attackers to use the Relative Path Overwrite (RPO) technique to perform.

**CVE-2020-8908**
                 A temp directory creation vulnerability exists in all versions of Guava, allowing an attacker with access to the machine to potentially access data in a temporary directory created by the Guava API com.google.common.io.Files.createTempDir(). By default, on unix-like systems, the created directory is world-readable (readable by an attacker with access to the system). The method in question has been marked @Deprecated in versions 30.0 and later and should not be used. For Android developers, we recommend choosing a temporary directory API provided by Android, such as context.getCacheDir(). For other Java developers, we recommend migrating to the Java 7 API java.nio.file.Files.createTempDirectory() which explicitly configures permissions of 700, or configuring the Java runtime\'s java.io.tmpdir system property to point to a location whose permissions are appropriately configured.

**CVE-2018-10237**
                 Unbounded memory allocation in Google Guava 11.0 through 24.x before 24.1.1 allows remote attackers to conduct denial of service attacks against servers that depend on this library and deserialize attacker-provided data, because the AtomicDoubleArray class (when serialized with Java serialization) and the CompoundOrdering class (when serialized with GWT serialization) perform eager allocation without appropriate checks on what a client has sent and whether the data size is reasonable.

**CVE-2020-13956**
                 Apache HttpClient versions prior to version 4.5.13 and 5.0.3 can misinterpret malformed authority component in request URIs passed to the library as java.net.URI object and pick the wrong target host for request execution.

### PR DESCRIPTION
**Pull request by settipalli2F51**
Fixes #23

### CODE REVIEW

**Review by swatij-ibm — CHANGES_REQUESTED** [review]
@settipalli2F51 

1. Does this even build ? 
Screenshot 2022-01-07 at 13 45 16
I don't see a 3.5.0 spring boot available https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-parent

2. I think we shouldn't be commenting out tests, I would prefer the tests to be handled and fixed. @hhellyer  what do you suggest/think on this ?

3. This repo does not have any automated Jenkins etc to check the workability of the new build. Last I recall, when vulnerabilities were fixed, before merging the changes into develop, a test install was run using a locally built image with the required changes. 
Has an installation test with these changes been tested ? 
If not then probably this needs to be included and done before the review and merge to validate the newly built image of sol installs fine.

**Review by clloydaccanto — CHANGES_REQUESTED** [review]
I agree with the comments left by @swatij-ibm

**Comment by settipalli2F51** [comment]
Hi @swatij-ibm,@clloydaccanto .

As per your review comments ,fixed the issue and created the below PR .
https://github.com/IBM/sol003-lifecycle-driver/pull/28
Please find the low local build screen shot .
image

**Comment by swatij-ibm** [comment]
Thanks for the update @settipalli2F51 .

As per the review comment provided above : https://github.com/IBM/sol003-lifecycle-driver/pull/28#pullrequestreview-846633636
> 2. I think we shouldn't be commenting out tests, I would prefer the tests to be handled and fixed. 

> 3. This repo does not have any automated Jenkins etc to check the workability of the new build. Last I recall, when vulnerabilities were fixed, before merging the changes into develop, a test install was run using a locally built image with the required changes.
    Has an installation test with these changes been tested ?
    If not then probably this needs to be included and done before the review and merge to validate the newly built image of sol installs fine.


Are you also looking to address the other requested changes ? 
I don't see the tests updated/fixed or any information about the installation test done for the new sol image ?

Please do address all the comments and also share details of the installation test being done, before sending out for re-review and merge.

**Comment by settipalli2F51** [comment]
Hi @swatij-ibm 

2.I think we shouldn't be commenting out tests, I would prefer the tests to be handled and fixed.
    ---> Now All Junit Test cases  are handled  by modifying the  Exception Handler .

3.This repo does not have any automated Jenkins etc to check the workability of the new build. Last I recall, when vulnerabilities were fixed, before merging the changes into develop, a test install was run using a locally built image with the required changes.
 -----> Done and Please find the latest installation  screen shot details .

image

Has an installation test with these changes been tested ?
--- >Yes ,  Ran the all test cases with out fail . 

image

**Review by swatij-ibm — APPROVED** [review]
Thanks for making the changes @settipalli2F51. 
LGTM 👍

File: pom.xml [hunk 1]
```diff
@@ -71,6 +71,11 @@
             <artifactId>commons-io</artifactId>
             <version>${commons-io.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>30.0-jre</version>
```
Comment (clloydaccanto) [inline pom.xml]: can this version number please be moved into the properties section where other version are managed

File: src/main/java/com/accantosystems/stratoss/vnfmdriver/web/etsi/ETSIExceptionHandlingControllerAdvice.java [hunk 2]
```diff
@@ -90,8 +90,8 @@ protected ProblemDetails handleNotImplementedException(HttpServletRequest req, N
     @ExceptionHandler(ResponseTypeNotAcceptableException.class)
     @ResponseStatus(HttpStatus.NOT_ACCEPTABLE)
     @ResponseBody
-    protected ProblemDetails handleResponseTypeNotAcceptableException(HttpServletRequest req, ResponseTypeNotAcceptableException cause) {
-        return defaultHandle("The requested response type was not acceptable", cause, HttpStatus.NOT_ACCEPTABLE);
+    protected ResponseEntity<String> handleResponseTypeNotAcceptableException(HttpServletRequest req, ResponseTypeNotAcceptableException cause) {
+        return new ResponseEntity<>("The requested response type was not acceptable", HttpStatus.NOT_ACCEPTABLE);
```
Comment (clloydaccanto) [inline src/main/java/com/accantosystems/stratoss/vnfmdriver/web/etsi/ETSIExceptionHandlingControllerAdvice.java]: sorry but I don't think this is the right fix. We can't just change the response type from a ProblemDetails structure (as expected by the ETSI spec) to plain string response, especially as it won't contain the cause as it did previously. Also by not calling defaultHandle we no longer have any of the logging that went on in that method

**Review by clloydaccanto — CHANGES_REQUESTED** [review]

**Comment by settipalli2F51** [comment]
we are planning to create new PR , So closing this PR .

**Pull request closed without merging**